### ISSUE TITLE
CSS custom property definition is excluded

### ISSUE THREAD

**Issue by cdoublev** [w3c/reffy#1567]
(Hi :metal:.) Could you please explain me the reason why the definition of [`--*`](https://drafts.csswg.org/css-variables-2/#propdef-) is [excluded](https://github.com/w3c/reffy/blob/212ffb2c41d21f8455019cd5e35f7be3ee29722a/src/postprocessing/csscomplete.js#L43) ?

I can easily work around this, and this definition is unlikely to change often, but I am just letting you know that I actually need it.

When parsing a CSS declaration, if its name starts with `--`, a CSS parser may need to resolve its `value` definition from the list of property definitions accepted in the context (CSS rule). Some contexts accept custom property declarations, others do not. When generating these lists of properties progamatically from the list of all property definitions, based on which are accepted or not in the context, the definition of `--*` should be included in the list of all property definitions.

A similar processing would also apply when resolving declaration values, to resolve the `computedValue` and `inherited` fields.

**Comment by tidoust** [comment]
Hello @cdoublev!

We excluded `--*` because it does not define an actual CSS **property**. It defines a **family of properties**. Conceptually, these are two different beasts. We'd need to separate `--*` from the other properties in the CSS extract of css-variables to avoid the risk that someone interprets `--*` literally as describing an existing property.

We could introduce a `families` key at the root of CSS extracts to hold such definitions. In practice, that's the only case of a family of properties that I'm aware of (if there are other cases, they went through the cracks!), so I'm not super enthusiastic. If you'd like to propose a pull request along with a test to validate the change, I could easily be persuaded though ;) (the only benefit I see is indeed the ability to track potential updates to this definition which, as you note, should be rare).

**Comment by cdoublev** [comment]
And `--*` is actually not a valid custom property name. This might be the main argument for not including it in `properties`.

About families of properties...

**TLDR:** what is a *family of properties* is unspecified and I have no strong opinion on which definition `properties` or `families` should be based on.

**CSS** defines `--*` as the generic name of the open-ended set of *custom **properties***. I could not find a formal definition of *standard **properties*** but spec authors often use this term for the closed-ended set of properties accepted in a style rule.

CSS also defines property sets that are accepted in a specific context. This [issue](https://github.com/w3c/csswg-drafts/issues/9453) is about consolidating these sets. I proposed to add a corresponding field in property definition tables.

Finally, CSS defines logical property groups, which have an influence on the order of declarations. There is a corresponding field in definition tables of properties that belong to a logical property group.

**MDN** uses different kinds of sets of properties in its navigation menu, under `Properties` (which does not include `--*`):

  - `-moz-*`, `-webkit-*`: property aliases
  - `align-*`, `color-*`, etc: properties targeting the same rendering aspect
  - `animation-*`, `border-*`, etc: shorthand properties and their subproperties

The third group can be considered a subgroup of the second.

Rules nested in [`@font-feature-values`](https://drafts.csswg.org/css-fonts-4/#font-feature-values-syntax) accept an open-ended set of descriptors. This is a simplified overview of the definition tree I use for parsing:

```js
const rules = {
  '@font-feature-values': {
    rules: {
      '@annotation': {
        descriptors: {
          '*': '<integer [0,∞]>',
        },
      },
    },
  },
  '@style': {
    properties: {
      '--*': '<declaration-value>?',
      // Standard properties
      ...require('webref-properties.js'),
    },
  },
}
```

`descriptors` and `properties` do not discriminate open/closed sets, therefore I do not think adding `families` would be usefull.

---

Do we close this issue with no change?
